src/discovered.rs

```diff
@@ -3,7 +3,12 @@ use biscuit::jwk::JWKSet;
 use biscuit::Empty;
 use reqwest::Client;
 use url::Url;
+#[cfg(feature = "uma2")]
+use crate::uma2::Uma2Provider;
 
+#[cfg(feature = "uma2")]
+pub struct Discovered(pub Config, pub Option<Config>);
```

Comment:
This structure signature just get my attention.

We should not change basic structure for optional feature.

Question here, is config for UMA2 a super set from OpenID config?

If so, we can just have Discovered<C=Config>(pub C);

And for Uma2 create special config Uma2Config which will have standard config with serde(flatten) attribute. This way can avoid duplicates.

Second option here is to just create Uma2Discovered in uma2 module and make it completely independent from openid part.

Do we need to query openid configuration at all? If no, I would think of second option as better option.

Reply:
UMA2 in itself is an extension of OpenID.
What I've noticed from the Keycloak implementation is that the UMA2 discovery endpoint has all the UMA2 additions but some of the OpenID optional entries are missing (so no to your last answer - but I haven't investigated the effects of the missing entries but let's assume it is very important).

My point -
I think the Discovered<C=Config>(pub C) would be doable - interrogate all the entries from OpenID and UMA2 and bail on conflicts.
More elegant is to make the Uma2Config struct - which I'll be implementing.

Reply:
I made a gist with the return of both endpoints:
https://gist.github.com/mvniekerk/bd9a11c32283a676a6bdd681187a36e6
When you do a diff on these using something like http://jsondiff.com/ you'll see the OpenID has lots more.
My suggestion - one Discovered<C=Config>(pub C) - but do interrogate the UMA2 endpoint as well and copy over the 3 extra endpoints over:  

```json
{"permission_endpoint": "",
    "policy_endpoint": "",
    "resource_registration_endpoint": ""}
```

Reply:
Implemented as by #eaf4465

Reply:
I think of Uma2Config + Uma2Discovered.

If discovery for openid would be successful, but next call to discover uma2 would fail because of network issue. As I see now, we would not notice this, just our logic have to adopt to this case.

And if we enable uma2 feature we cannot do discovery without uma2, there is no explicit way to tell, that exactly we want.

With separate Uma2Discovery we do not need to deal with it and we can also fail if there is no UMA2 support.

Open question for me, if it is uma2 - yes, we have less fields for different discoveries, but probably we do not need those openid fields either. Or another possibility is not complete implementation / configuration in Keycloak. Can you check, if you just do not do discovery for openid and leave only uma2 configuration, would it work for you?

Reply:
It works for me. Still thinking how to refactor in that we can reuse the Discovered / Config code.

Reply:
We have similar thing with CustomClaims. I will take a look today also...

Reply:
Ok so https://github.com/kilork/openid/pull/4/commits/16fce3aeeac9c3b1288371cd23ebbd56806d7e0b uses UMA2 discovery as its only discovery, and it works for me and my flow (obviously).